central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -0,0 +1,398 @@
+package datastore
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/stackrox/rox/central/metrics"
+	processIndicatorStore "github.com/stackrox/rox/central/processindicator/datastore"
+	"github.com/stackrox/rox/central/processlisteningonport/store/postgres"
+	"github.com/stackrox/rox/central/role/resources"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/search"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+type datastoreImpl struct {
+	storage            postgres.Store
+	indicatorDataStore processIndicatorStore.DataStore
+}
+
+var (
+	plopSAC = sac.ForResource(resources.DeploymentExtension)
+	log     = logging.LoggerForModule()
+)
+
+func newDatastoreImpl(
+	storage postgres.Store,
+	indicatorDataStore processIndicatorStore.DataStore,
+) *datastoreImpl {
+	return &datastoreImpl{
+		storage:            storage,
+		indicatorDataStore: indicatorDataStore,
+	}
+}
+
+func (ds *datastoreImpl) AddProcessListeningOnPort(
+	ctx context.Context,
+	portProcesses ...*storage.ProcessListeningOnPortFromSensor,
+) error {
+	defer metrics.SetDatastoreFunctionDuration(
+		time.Now(),
+		"ProcessListeningOnPort",
+		"AddProcessListeningOnPort",
+	)
+
+	if !env.PostgresDatastoreEnabled.BooleanSetting() {
+		// PLOP is a Postgres-only feature, do nothing.
+		log.Warnf("Tried to add PLOP not on Postgres, ignore: %+v", portProcesses)
+		return nil
+	}
+
+	if ok, err := plopSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	portProcesses = normalizePLOPs(portProcesses)
+
+	// XXX: The next two calls, fetchIndicators and fetchExistingPLOPs, have to
+	// be done in a single join query fetching both ProcessIndicator and needed
+	// bits from PLOP.
+	indicatorsMap, indicatorIds, err := ds.fetchIndicators(ctx, portProcesses...)
+	if err != nil {
+		return err
+	}
+
+	existingPLOPMap, err := ds.fetchExistingPLOPs(ctx, indicatorIds, portProcesses...)
+	if err != nil {
+		return err
+	}
+
+	plopObjects := make([]*storage.ProcessListeningOnPortStorage, len(portProcesses))
+	for i, val := range portProcesses {
+		indicatorID := ""
+		var processInfo *storage.ProcessIndicatorUniqueKey
+
+		key := getProcesUniqueKey(val)
+
+		if indicator, ok := indicatorsMap[key]; ok {
+			indicatorID = indicator.GetId()
+			log.Debugf("Got indicator %s: %+v", indicatorID, indicator)
+		} else {
+			// XXX: Create a metric for this
+			log.Warnf("Found no matching indicators for %s", key)
+			processInfo = val.Process
+		}
+
+		plopKey := fmt.Sprintf("%d %d %s",
```

Comment:
nit:  This likely doesn't matter at all, but if we were to ever look at these keys visually it might be better to have a delimiter other than a space.  a `-` or `_` would be more obvious that it is a multiple items mashed together into 1 than a space does.

Reply:
We now have various functions to get the various keys and all keys are obtained by calling those functions. All of them use "_" as the delimiter.